pkg/chain/ethereum/ethereum.go

```diff
@@ -2,458 +2,93 @@ package ethereum
 
 import (
 	"context"
-	"encoding/binary"
 	"fmt"
-	"math/big"
-	"math/rand"
+	"github.com/ethereum/go-ethereum/accounts/keystore"
+	"github.com/keep-network/keep-common/pkg/chain/ethlike"
 
 	"github.com/ipfs/go-log"
-
-	"github.com/ethereum/go-ethereum/common"
-	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/keep-network/keep-common/pkg/chain/ethereum"
 	"github.com/keep-network/keep-common/pkg/chain/ethereum/ethutil"
-	relayChain "github.com/keep-network/keep-core/pkg/beacon/relay/chain"
-	"github.com/keep-network/keep-core/pkg/beacon/relay/event"
-	"github.com/keep-network/keep-core/pkg/chain"
-	"github.com/keep-network/keep-core/pkg/gen/async"
-	"github.com/keep-network/keep-core/pkg/operator"
-	"github.com/keep-network/keep-core/pkg/subscription"
+	"github.com/keep-network/keep-common/pkg/rate"
 )
 
 var logger = log.Logger("keep-chain-ethereum")
 
-// ThresholdRelay converts from ethereumChain to beacon.ChainInterface.
-func (ec *ethereumChain) ThresholdRelay() relayChain.Interface {
-	return ec
-}
-
-func (ec *ethereumChain) GetKeys() (
-	*operator.PrivateKey,
-	*operator.PublicKey,
-	error,
-) {
-	privateKey, publicKey, err := ChainPrivateKeyToOperatorKeyPair(
-		ec.accountKey.PrivateKey,
-	)
+// Chain represents a base, non-application-specific chain handle. It
+// provides the implementation of generic features like balance monitor,
+// block counter and similar.
+type Chain struct {
+	key          *keystore.Key
+	client       ethutil.EthereumClient
+	blockCounter *ethlike.BlockCounter
+}
+
+// NewChain construct a new instance of the Ethereum chain handle.
+func NewChain(
+	ctx context.Context,
+	config *ethereum.Config,
+	client ethutil.EthereumClient,
```

Comment:
This should probably be `*ethclient.Client,` from `github.com/ethereum/go-ethereum/ethclient`. We need to obtain `ChainID` with a call to `chainID, err := client.ChainID(context.Background())` and this is not possible with `ethutil.EthereumClient`.

Reply:
I did this change in #3067 at commit https://github.com/keep-network/keep-core/pull/3067/commits/012e36247d967791a7f8bbd73ce85f68ff7b5010. Started building on a top of this PR.